sklearn/cluster/_kmeans.py

```diff
@@ -2144,17 +2144,27 @@ def fit(self, X, y=None, sample_weight=None):
         self._n_since_last_reassign = 0
 
         n_steps = (self.max_iter * n_samples) // self._batch_size
-
+        normalized_sample_weight = sample_weight / np.sum(sample_weight)
+        unit_sample_weight = np.ones_like(sample_weight, shape=(self._batch_size,))
         with _get_threadpool_controller().limit(limits=1, user_api="blas"):
             # Perform the iterative optimization until convergence
             for i in range(n_steps):
                 # Sample a minibatch from the full dataset
-                minibatch_indices = random_state.randint(0, n_samples, self._batch_size)
+                minibatch_indices = random_state.choice(
+                    n_samples,
+                    self._batch_size,
+                    p=normalized_sample_weight,
+                    replace=True,
+                )
 
                 # Perform the actual update step on the minibatch data
+                # Note, I am not sure how sample weights are used here
+                # So left it in, it seems like the weight sums are updated using
+                # sample weights so need some help here to understand the
+                # _minibatch_update_dense/sparse code
```

Comment:
```suggestion
                # Note: since the sampling of the minibatch is sample_weight aware,
                # we pass fixed unit weights to the `_mini_batch_step` call to avoid
                # accounting for the weights twice. Also note that `_mini_batch_step`
                # can be called with non-unit weights when the caller constructs
                # the batches explicitly by calling the public `partial_fit` method
                # instead.
```